File: Sources/YTools/Services/ResultAggregator.swift
import Foundation

struct AggregatedResults {
    let results: [LauncherResult]
    let selectedIndex: Int
}

/// Owns result ranking and privacy-preserving usage learning.
@MainActor
final class ResultAggregator {
    private let usage: UsageRankingStore

    init(usage: UsageRankingStore = UsageRankingStore()) {
        self.usage = usage
    }

    func aggregate(
        background: [LauncherResult],
        spotlight: [LauncherResult],
        query: String,
        previousResults: [LauncherResult],
        selectedIndex: Int
    ) -> AggregatedResults {
        let selectedID = previousResults.indices.contains(selectedIndex)
            ? previousResults[selectedIndex].id
            : nil
        let results = (background + spotlight)
            .map { result in
                result.withScore(result.score + usage.boost(for: result.id, query: query))
            }
            .sorted {
                if $0.score != $1.score { return $0.score > $1.score }
                let leftTieCount = $0.moduleID == "applications" ? usage.count(for: $0.id) : 0
                let rightTieCount = $1.moduleID == "applications" ? usage.count(for: $1.id) : 0
                if leftTieCount != rightTieCount { return leftTieCount > rightTieCount }
                return $0.title.localizedStandardCompare($1.title) == .orderedAscending
            }
        let nextIndex: Int
        if let selectedID, let retained = results.firstIndex(where: { $0.id == selectedID }) {
            nextIndex = retained
        } else {
            nextIndex = results.isEmpty ? 0 : min(selectedIndex, results.count - 1)
        }
        return AggregatedResults(results: results, selectedIndex: nextIndex)
    }

    func record(_ result: LauncherResult, query: String) {
        usage.record(result.id, query: query)
    }

    func clearLearning() {
        usage.clear()
    }
}

File: src/YTools.Windows/Services/ResultAggregator.cs
using YTools.Models;
using YTools.ModuleKit;
using YTools.Services.Storage;

namespace YTools.Services;

public sealed record AggregatedResults(IReadOnlyList<LauncherResult> Results, int SelectedIndex);

/// <summary>Owns result ranking and privacy-preserving usage learning.</summary>
public sealed class ResultAggregator
{
    private readonly UsageRankingStore _usage;

    public ResultAggregator(UsageRankingStore? usage = null)
    {
        _usage = usage ?? new UsageRankingStore();
    }

    public AggregatedResults Aggregate(
        IReadOnlyList<LauncherResult> background,
        IReadOnlyList<LauncherResult> fileResults,
        string query,
        IReadOnlyList<LauncherResult> previousResults,
        int selectedIndex)
    {
        var selectedId = previousResults.Count > selectedIndex && selectedIndex >= 0
            ? previousResults[selectedIndex].Id
            : null;
        var results = background.Concat(fileResults)
            .Select(result => result.WithScore(result.Score + _usage.Boost(result.Id, query)))
            .OrderByDescending(result => result.Score)
            .ThenBy(result => result.Title, StringComparer.OrdinalIgnoreCase)
            .ToList();

        int nextIndex;
        if (selectedId is not null
            && results.FindIndex(result => result.Id == selectedId) is { } retained
            && retained >= 0)
        {
            nextIndex = retained;
        }
        else
        {
            nextIndex = results.Count == 0 ? 0 : Math.Min(selectedIndex, results.Count - 1);
        }

        return new AggregatedResults(results, nextIndex);
    }

    public void Record(LauncherResult result, string query)
    {
        _usage.Record(result.Id, query);
    }

    public void ClearLearning()
    {
        _usage.Clear();
    }
}
